server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -142,6 +144,13 @@ public synchronized void updateSegments(final SegmentInfos infos, long seqNo) th
     public void commitSegmentInfos() throws IOException {
         // TODO: This method should wait for replication events to finalize.
         final SegmentInfos latestSegmentInfos = getLatestSegmentInfos();
+        /*
+         This is a workaround solution which decreases the chances of conflict on replica nodes when same file is copied
+         from two different primaries during failover. Increasing counter helps in avoiding this conflict as counter is
+         used to generate new segment file names. The ideal solution is to identify the counter from previous primary.
+         */
+        latestSegmentInfos.counter = latestSegmentInfos.counter + SI_COUNTER_INCREMENT;
```

Comment:
I'm not opposed to this as a "best-effort" approach but it would be great if we could somehow fetch and increment from the previous primary's counter instead of this replica's latest infos.  latestSegmentInfos on the replica here could still potentially be more than 10 generations behind and get promoted as primary.

I'm not sure exactly how we'd do that though with any accuracy. 

I'm ok with this as best-effort, because we can still fail & recover the replicas that have conflicts as needed and attempt to select the furthest ahead during promotion.

Reply:
>I'm not sure exactly how we'd do that though with any accuracy.
I added few alternatives in the description. 

This solution works well and essentially prevents file conflicts on nodes, which is in general helpful not just for linked bug but overall working of segment replication.